### PR TITLE
Add caching during protobuf generation

### PR DESCRIPTION
**Pull request by agrawal-siddharth**

### CODE REVIEW

File: spark-bigquery-connector-common/src/test/java/com/google/cloud/spark/bigquery/ProtobufUtilsTest.java [hunk 1]
```diff
@@ -115,43 +115,51 @@ public void testBigQueryToProtoSchema() throws Exception {
   public void testSparkStructRowToDynamicMessage() throws Exception {
     StructType schema = new StructType().add(SPARK_NESTED_STRUCT_FIELD);
     Descriptors.Descriptor schemaDescriptor = toDescriptor(schema);
-    Message converted = buildSingleRowMessage(schema, schemaDescriptor, STRUCT_INTERNAL_ROW);
+    Message converted =
+        buildSingleRowMessage(
+            schema,
+            schemaDescriptor,
+            STRUCT_INTERNAL_ROW,
+            /*fieldIndexToEntryMap*/ null,
```
Comment (vishalkarve15) [inline spark-bigquery-connector-common/src/test/java/com/google/cloud/spark/bigquery/ProtobufUtilsTest.java:123]: Can we add a test to verify cached entries?

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/BigQueryConnectorUtils.java [hunk 2]
```diff
@@ -52,4 +57,32 @@ static SparkBigQueryPushdown createSparkBigQueryPushdown() {
...
             SparkBigQueryPushdown.class.getName(), sparkVersion));
   }
+
+  public static final class ProtobufSchemaFieldCacheEntry {
+    DataType sparkType;
+    boolean nullable;
+    Descriptors.Descriptor nestedTypeDescriptor;
+    Descriptors.FieldDescriptor fieldDescriptor;
+    Optional<TypeConverter> typeConverterOptional;
+    Optional<SupportedCustomDataType> customDataTypeOptional;
```
Comment (vishalkarve15) [inline spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/BigQueryConnectorUtils.java:67]: Please use getters + (setters / constuctors).

Comment (vishalkarve15) [reply spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/BigQueryConnectorUtils.java:67]: Also, this class should be moved to ProtobufUtils.

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/ProtobufUtils.java [hunk 3]
```diff
@@ -298,27 +306,40 @@ public static ProtoRows toProtoRows(StructType sparkSchema, InternalRow[] rows)
...
-      StructField sparkField = schema.fields()[fieldIndex];
-      DataType sparkType = sparkField.dataType();
-
-      Object sparkValue = row.get(fieldIndex, sparkType);
-      boolean nullable = sparkField.nullable();
-      Descriptors.Descriptor nestedTypeDescriptor =
-          schemaDescriptor.findNestedTypeByName(RESERVED_NESTED_TYPE_NAME + (protoFieldNumber));
+      int fieldIndexCopy = fieldIndex;
+      BigQueryConnectorUtils.ProtobufSchemaFieldCacheEntry protobufSchemaFieldEntry =
+          (fieldIndexToEntryMap != null)
```
[the remark this answers is not in the document]
Comment (agrawal-siddharth) [reply spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/ProtobufUtils.java]: Done

File: spark-bigquery-connector-common/src/test/java/com/google/cloud/spark/bigquery/ProtobufUtilsTest.java (hunk 1 above)
Comment (agrawal-siddharth) [reply spark-bigquery-connector-common/src/test/java/com/google/cloud/spark/bigquery/ProtobufUtilsTest.java]: I added an additional test here. Previously I was relying on the call to toProtoRows() with multiple input rows to test it.

File: spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/BigQueryConnectorUtils.java (hunk 2 above)
Comment (agrawal-siddharth) [reply spark-bigquery-connector-common/src/main/java/com/google/cloud/spark/bigquery/BigQueryConnectorUtils.java]: Done

**Review by vishalkarve15 — APPROVED** [review]

**Pull request merged**